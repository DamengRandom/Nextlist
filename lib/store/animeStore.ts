import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { AnimeProps } from '../../types/anilist';

interface AnimeStates {
  type: string | undefined;
  sort: string | undefined;
  page: number;
  searchName: string | undefined;
  anime: AnimeProps | null;
  setType: (type: string) => void;
  setSort: (sort: string) => void;
  setPage: (page: number) => void;
  setSearchName: (searchName: string) => void;
  reset: () => void;
}

export const useAnimeStore = create<AnimeStates>()(
  persist(
    (set) => ({
      type: undefined,
      sort: undefined,
      page: 1,
      searchName: undefined,
      anime: null,
      setType: (type) => set({ type }),
      setSort: (sort) => set({ sort }),
      setPage: (page) => set({ page }),
      setSearchName: (searchName) => set({ searchName }),
      reset: () => set({ page: 1, searchName: undefined }),
    }),
    { name: 'anime-store' }
  )
);
