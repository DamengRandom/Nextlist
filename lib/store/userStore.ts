import { create } from 'zustand';
import { persist } from 'zustand/middleware';

interface UserState {
  username: string;
  jobTitle: string;
  setUsername: (username: string) => void;
  setJobTitle: (jobTitle: string) => void;
}

export const useUserStore = create<UserState>()(
  persist(
    (set) => ({
      username: '',
      jobTitle: '',
      setUsername: (username) => set({ username }),
      setJobTitle: (jobTitle) => set({ jobTitle }),
    }),
    { name: 'user-store' }
  )
);