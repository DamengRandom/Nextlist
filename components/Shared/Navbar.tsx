import { Flex, Box, Button, Heading, Spacer } from "@chakra-ui/react";
import { useAnimeStore } from "../../lib/store/animeStore";
import { useNavigation } from "../../hooks/useNavigation";

type NavbarProps = {
  handleEdit: () => void;
  username: string | null;
}

export default function Navbar({ handleEdit, username }: NavbarProps) {
  const { reset } = useAnimeStore();
  const { navigateTo } = useNavigation();

  const handleClick = () => {
    navigateTo('/');
    reset();
  }

  return (
    <Flex as="nav" align="center" justify="space-between" padding={4} bgGradient="linear(to-r, teal.500, blue.500)" color="white" boxShadow="md">
      <Box>
        <Heading size="lg" textShadow="0 1px 2px rgba(0,0,0,0.5)" cursor="pointer" onClick={handleClick}>Nextlist</Heading>
      </Box>
      <Spacer />
      <Box>
        <Button
          aria-pressed={!!username}
          onClick={handleEdit}
          variant="outline"
          color="white"
          border="2px solid transparent"
          backgroundClip="padding-box"
          position="relative"
          _before={{
            content: '""',
            position: "absolute",
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            zIndex: -1,
            margin: "-2px",
            borderRadius: "inherit",
            background: "linear-gradient(45deg, var(--chakra-colors-pink-500), var(--chakra-colors-blue-400))",
            opacity: 0,
            transition: "opacity 0.3s",
          }}
          _hover={{
            _before: {
              opacity: 1,
            },
            transform: "scale(1.02)",
          }}
        >
          {username ? `Hello ${username}` : "Sign In"}
        </Button>
      </Box>
    </Flex>
  );
}