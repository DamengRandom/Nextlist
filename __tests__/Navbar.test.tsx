import { render, screen, fireEvent } from '@testing-library/react';
import Navbar from '../components/Shared/Navbar';
import { useRouter } from 'next/navigation';
import { useAnimeStore } from '../lib/store/animeStore';
import { ChakraProvider } from '@chakra-ui/react';

// Mock the next/navigation module
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

// Mock the anime store
jest.mock('../lib/store/animeStore', () => ({
  useAnimeStore: jest.fn(),
}));

describe('Navbar component', () => {
  const mockPush = jest.fn();
  const mockReset = jest.fn();
  const mockHandleEdit = jest.fn();

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });
    (useAnimeStore as unknown as jest.Mock).mockReturnValue({
      reset: mockReset,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderNavbar = (props = {}) => {
    return render(
      <ChakraProvider>
        <Navbar handleEdit={mockHandleEdit} username="testuser" {...props} />
      </ChakraProvider>
    );
  };

  it('renders correctly with username', () => {
    renderNavbar({ username: "testuser" });
    
    expect(screen.getByText('Nextlist')).toBeInTheDocument();
    expect(screen.getByText('Hello testuser')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'true');
  });

  it('renders correctly without username', () => {
    renderNavbar({ username: null });
    
    expect(screen.getByText('Nextlist')).toBeInTheDocument();
    expect(screen.getByText('Sign In')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'false');
  });

  it('calls handleEdit when button is clicked', () => {
    renderNavbar({ username: "testuser" });
    
    fireEvent.click(screen.getByRole('button'));
    expect(mockHandleEdit).toHaveBeenCalledTimes(1);
  });

  it('navigates to home and resets store when title is clicked', () => {
    renderNavbar({ username: "testuser" });
    
    fireEvent.click(screen.getByText('Nextlist'));
    expect(mockPush).toHaveBeenCalledWith('/');
    expect(mockReset).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot with username', () => {
    const { asFragment } = renderNavbar({ username: "testuser" });
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot without username', () => {
    const { asFragment } = renderNavbar({ username: null });
    expect(asFragment()).toMatchSnapshot();
  });
});