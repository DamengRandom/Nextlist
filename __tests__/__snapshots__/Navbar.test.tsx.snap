// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component matches snapshot with username 1`] = `
<DocumentFragment>
  <nav
    class="css-6qkd65"
  >
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-5yayc3"
      >
        Nextlist
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      class="css-0"
    >
      <button
        aria-pressed="true"
        class="chakra-button css-14j3nze"
        type="button"
      >
        Hello testuser
      </button>
    </div>
  </nav>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Navbar component matches snapshot without username 1`] = `
<DocumentFragment>
  <nav
    class="css-6qkd65"
  >
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-5yayc3"
      >
        Nextlist
      </h2>
    </div>
    <div
      class="css-17xejub"
    />
    <div
      class="css-0"
    >
      <button
        aria-pressed="false"
        class="chakra-button css-14j3nze"
        type="button"
      >
        Sign In
      </button>
    </div>
  </nav>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
